src/test/java/com/aws/greengrass/authorization/WildcardVariableTrieTest.java

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.authorization;
+
+import com.aws.greengrass.testcommons.testutilities.GGExtension;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+@ExtendWith(GGExtension.class)
+class WildcardVariableTrieTest {
+    @Test
+    void testGlobWildcardMatching() {
+        WildcardVariableTrie rt = new WildcardVariableTrie();
+        rt.add("abc*xyz*");
+        rt.add("*/def");
+        assertTrue(rt.matches("abc1234asdxyz456", true));
+        assertTrue(rt.matches("abcdxyz", true));
+        assertTrue(rt.matches("/def", true));
+        assertTrue(rt.matches("12345/def", true));
+        assertTrue(rt.matches(null, true));
+        assertFalse(rt.matches("abc123xyz456/89", true));
+        assertFalse(rt.matches("2/3/def", true));
+        assertFalse(rt.matches("", true));
+
+        // Only '*' should match all resources
+        WildcardVariableTrie rt1 = new WildcardVariableTrie();
+        rt1.add("*");
+        rt1.add("*wer");
+        assertTrue(rt1.matches("9999/88", true));
+
+        WildcardVariableTrie rt2 = new WildcardVariableTrie();
+        rt2.add("a/*/*/d");
+        rt2.add("**/abc");
+        assertTrue(rt2.matches("a/12/34/d", true));
+        assertFalse(rt2.matches("a/2/3/4/d", true));
+        assertTrue(rt2.matches("78/abc", true));
+        assertFalse(rt2.matches("7/8/abc", true));
+    }
+
+    @Test
+    void testMQTTMultilevelWildcardMatching() {
+        //test MQTT wildcard usages according to MQTT V5.0 spec (https://docs.oasis-open.org/mqtt/mqtt/v5.0/mqtt-v5.0.html)
+        // Test Valid usages
+        WildcardVariableTrie rt = new WildcardVariableTrie();
+        rt.add("abc/#");
+        rt.add("123/#/xyz/#");
+        rt.add("#/zzz/45");
```

Comment:
I'd probably add these one at a time and test the appropriate cases. As it is now, we don't really know which of these several options is causing the match to succeed. If the implementation is incorrect, then the on starting with `#` may be allowing all of them through; I trust this isn't happening, but let's make sure that's impossible.